### PR TITLE
[Whisper] Load customized MLX model & Quantization

### PR DESCRIPTION
**Pull request by bofenghuang**
Hi @awni 👋,

This PR introduces a feature to load customized MLX whisper models and integrates a conversion script for transforming the PyTorch checkpoint of OpenAI format into MLX. I believe this will be helpful for the community to easily convert and load other whisper models fine-tuned for low-resource languages in MLX.

Additionally, I've incorporated quantization into both the model conversion and loading processes.

It's working on my side, but might need additional review for modifications, doc, or testing. Thanks in advance!

### CODE REVIEW

**Comment by awni** [comment]
Wow that's all awesome!! Thanks for adding! Will review soon!

**Review by awni — CHANGES_REQUESTED** [review]
This is nicely done and super helpful! Thank you!!

I left a few minor comments, let me know when it's ready and we can merge!

Also I will test the quantization on my end to make sure it works. Did you notice any speedup with the larger models? (There may be quite a few overheads in this example as it closely follows the OpenAI one which looked pretty suboptimal with all the pre/post processing in python.)

File: whisper/convert.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from whisper.torch_whisper import ModelDimensions
+from whisper.whisper import Whisper
+
+MODEL_DTYPES = {"float16", "float32"}
+
+def quantize(weights, config, dtype, args):
+    quantized_config = copy.deepcopy(config)
+
+    # Load the model:
+    model = Whisper(ModelDimensions(**config), dtype)
```
[the remark this answers is not in the document]
Comment (bofenghuang) [reply whisper/convert.py:24]: Yeah, I was thinking of having an option to keep some non-linear layers in FP32. It seems unnecessary if we're already doing quantization.

File: whisper/convert.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        "group_size": args.q_group_size,
+        "bits": args.q_bits,
+    }
+    quantized_weights = dict(tree_flatten(model.parameters()))
+
+    return quantized_weights, quantized_config
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Convert Mistral weights to MLX.")
```
[the remark this answers is not in the document]
Comment (bofenghuang) [reply whisper/convert.py:42]: oops

File: whisper/whisper/load_models.py [hunk 3]
```diff
@@ -191,8 +194,51 @@ def convert_rblock(model, rules):
...
+        sharded_weights_glob = str(model_path / "weights.*.npz")
+        weight_files = glob.glob(sharded_weights_glob)
+        print(f"[INFO] Loading model from {sharded_weights_glob}")
+
+        if len(weight_files) == 0:
+            raise FileNotFoundError("No weights found in {}".format(model_path))
+
+        weights = {}
+        for wf in weight_files:
+            weights.update(mx.load(wf).items())
```
[the remark this answers is not in the document]
Comment (bofenghuang) [reply whisper/whisper/load_models.py:226]: agree !

File: whisper/whisper/load_models.py [hunk 4]
```diff
@@ -191,8 +194,51 @@ def convert_rblock(model, rules):
...
 def load_model(
-    name: str,
+    name_or_path: str,
     download_root: str = None,
     dtype: mx.Dtype = mx.float32,
 ) -> whisper.Whisper:
-    return torch_to_mlx(load_torch_model(name, download_root), dtype)
+    if name_or_path in _MODELS:
+        print(f"[INFO] Loading and converting {name_or_path} model")
+        return torch_to_mlx(load_torch_model(name_or_path, download_root), dtype)
```
Comment (bofenghuang) [inline whisper/whisper/load_models.py:203]: Do you think it would be better to move this and all conversion utilities (`_download`, `load_torch_model`, `convert`, `torch_to_mlx`) into convert.py, to separate conversion and loading, as has been done for other models?

File: whisper/whisper/load_models.py [hunk 5]
```diff
@@ -191,8 +194,51 @@ def convert_rblock(model, rules):
...
+        weights = {}
+        for wf in weight_files:
+            weights.update(mx.load(wf).items())
+
+    with open(model_path / "config.json", "r") as f:
+        config = json.loads(f.read())
+        config.pop("model_type", None)
+        quantization = config.pop("quantization", None)
+
+    model_args = torch_whisper.ModelDimensions(**config)
```
[the remark this answers is not in the document]
Comment (bofenghuang) [reply whisper/whisper/load_models.py:233]: didn't see it. thanks!

**Comment by bofenghuang** [comment]
Thanks for the review! Unfortunately, I didn't see any speed improvement from tests run on my M1 Mac. In fact, it's a bit slower. The benefit comes from memory usage. I'm not entirely sure what happened here, but I'm looking into it.

```
Feature time 0.037

Model: MODELS/LARGE
Model forward time 1.197
Decode time 2.569
Everything time 3.634

--------------------------------------------------

Feature time 0.036

Model: MODELS/LARGE_QUANTIZED_8BITS
Model forward time 1.506
Decode time 2.965
Everything time 4.285

--------------------------------------------------


Feature time 0.036

Model: MODELS/LARGE_QUANTIZED_4BITS
Model forward time 1.536
Decode time 2.913
Everything time 4.262
```

File: whisper/convert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awni) [reply whisper/convert.py]: > Yeah, I was thinking of having an option to keep some non-linear layers in FP32

The dtype is used to construct the mask and the sinusoidal position encodings, (both of which are not parameters). Those don't even get saved / returned in the `model.paramters()`, so this really doesn't have any affect. All the other params are overwritten by the input weights when you do `model.udpate(...)` in the line below.

File: whisper/whisper/load_models.py (hunk 4 above)
Comment (awni) [reply whisper/whisper/load_models.py:203]: Yea, I think that would probably be cleaner. (But I'm ok merging this PR as is as well). I'll leave it to you to decide on that.

**Review by awni — APPROVED** [review]
🚀  This is great, thanks for the additions!!

@bofenghuang let me know if you want to make any other changes before I mere it.

File: whisper/convert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bofenghuang) [reply whisper/convert.py]: I was thinking about the CNN layers since the weights are converted using `tree_map(lambda p: p.astype(dtype), params)` before being passed to models. But I get your point, and also believe it's basically the same thing.

File: whisper/whisper/load_models.py (hunk 4 above)
Comment (bofenghuang) [reply whisper/whisper/load_models.py:203]: let me make it cleaner, can also encourage users to convert the model once and for all, and upload more models to Hugging Face :)

Comment (awni) [reply whisper/whisper/load_models.py:203]: Awesome!! Would love to see some models on Hugging Face. The MLX community would be a great place to put them: https://huggingface.co/mlx-community

**Comment by bofenghuang** [comment]
@awni just finished and updated the test and benchmark to align with the change. thanks for the review!

File: whisper/README.md [hunk 6]
```diff
@@ -19,6 +19,27 @@ Install [`ffmpeg`](https://ffmpeg.org/):
...
 ```
 
+Next, download the Whisper PyTorch checkpoint and convert the weights to MLX format:
+
+```
+# Take the "tiny" model as an example. Note that you can also convert a local PyTorch checkpoint in OpenAI's format.
+python convert.py --torch-name-or-path tiny --mlx-path mlx_models/tiny
+```
+
+To generate a 4-bit quantized model, use ``-q`` for a full list of options:
```
Comment (awni) [inline whisper/README.md:29]: -q should have single back ticks.

Also put a period after it and start a new sentence at the `For ...`

File: whisper/README.md [hunk 7]
```diff
@@ -19,6 +19,27 @@ Install [`ffmpeg`](https://ffmpeg.org/):
 brew install ffmpeg
 ```
 
+Next, download the Whisper PyTorch checkpoint and convert the weights to MLX format:
+
+```
+# Take the "tiny" model as an example. Note that you can also convert a local PyTorch checkpoint in OpenAI's format.
```
Comment (awni) [inline whisper/README.md:25]: Let's move this comment out as text:

convert the weights to the MLX format. For exampl, to convert the `tiny` model use:

```
python convert.py
```

Note you can also convert a local PyTorch checkpoint which is in the original OpenAI format.

File: whisper/README.md (hunk 6 above)
Comment (bofenghuang) [reply whisper/README.md:29]: oops

File: whisper/README.md (hunk 7 above)
Comment (bofenghuang) [reply whisper/README.md]: yes 👍

**Review by awni — APPROVED** [review]
🚀  awesome thanks!!
